gk/fib.c

```diff
@@ -0,0 +1,1410 @@
+		return NULL;
+	}
+
+	if (action == GK_FWD_GATEWAY_FRONT_NET)
+		iface = &gk_conf->net->front;
+	else if (likely(action == GK_FWD_GATEWAY_BACK_NET))
+		iface = &gk_conf->net->back;
+	else {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since it has invalid action %d.\n",
+			action);
+		return NULL;
+	}
+
+	/* Find the neighbor FIB entry for this gateway. */
+	neigh_fib = find_fib_entry_for_neighbor_locked(
+		gw_addr, action, gk_conf);
+	if (neigh_fib == NULL)
+		return NULL;
+
+	/* Find the Ethernet cached header entry for this gateway. */
+	neigh_ht = &neigh_fib->u.neigh;
+	eth_cache = neigh_get_ether_cache_locked(neigh_ht, gw_addr, iface);
+	if (eth_cache == NULL)
+		return NULL;
+
+	/* Find an empty FIB entry for the Gateway. */
+	fib_id = get_empty_fib_id(ip_prefix->addr.proto, gk_conf);
+	if (fib_id < 0) {
+		if (eth_cache->ref_cnt == 0)
+			neigh_del_ether_cache(neigh_ht, &gw_addr->ip);
+		return NULL;
+	}
+
+	if (ip_prefix->addr.proto == ETHER_TYPE_IPv4)
+		gw_fib = &ltbl->fib_tbl[fib_id];
+	else
+		gw_fib = &ltbl->fib_tbl6[fib_id];
+
+	/* Fills up the Gateway FIB entry for the IP prefix. */
+	gw_fib->action = action;
+	gw_fib->u.gateway.eth_cache = eth_cache;
+
+	ret = lpm_add_route(&ip_prefix->addr, ip_prefix->len, fib_id, ltbl);
+	if (ret < 0) {
+		ether_cache_put(neigh_fib, action, eth_cache, gw_addr, gk_conf);
+		initialize_fib_entry(gw_fib);
+		return NULL;
+	}
+
+	return gw_fib;
+}
+
+static struct gk_fib *
+init_grantor_fib_locked(
+	struct ip_prefix *prefix, struct ipaddr *gt_addr,
+	struct ipaddr *gw_addr, struct gk_config *gk_conf)
+{
+	int ret, fib_id;
+	struct gk_fib *gt_fib;
+	struct ether_cache *eth_cache;
+	struct neighbor_hash_table *neigh_ht = NULL;
+	struct gatekeeper_if *iface = &gk_conf->net->back;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	/* Find the neighbor FIB entry for this gateway. */
+	struct gk_fib *neigh_fib = find_fib_entry_for_neighbor_locked(
+		gw_addr, GK_FWD_GATEWAY_BACK_NET, gk_conf);
+	if (neigh_fib == NULL)
+		return NULL;
+
+	/* Find the Ethernet cached header entry for this gateway. */
+	neigh_ht = &neigh_fib->u.neigh;
+	eth_cache = neigh_get_ether_cache_locked(neigh_ht, gw_addr, iface);
+	if (eth_cache == NULL)
+		return NULL;
+
+	fib_id = get_empty_fib_id(prefix->addr.proto, gk_conf);
+	if (fib_id < 0)
+		return NULL;
+
+	if (prefix->addr.proto == ETHER_TYPE_IPv4)
+		gt_fib = &ltbl->fib_tbl[fib_id];
+	else
+		gt_fib = &ltbl->fib_tbl6[fib_id];
+
+	gt_fib->action = GK_FWD_GRANTOR;
+	rte_memcpy(&gt_fib->u.grantor.gt_addr,
+		gt_addr, sizeof(gt_fib->u.grantor.gt_addr));
+	gt_fib->u.grantor.eth_cache = eth_cache;
+
+	ret = lpm_add_route(&prefix->addr, prefix->len, fib_id, ltbl);
+	if (ret < 0) {
+		ether_cache_put(neigh_fib,
+			GK_FWD_GATEWAY_BACK_NET, eth_cache, gw_addr, gk_conf);
+		initialize_fib_entry(gt_fib);
+		return NULL;
```

Comment:
1. Like in `init_gateway_fib_locked()`, use stacked `goto` to handle errors.

2. Don't forget to `goto` a label that calls `ether_cache_put()` when `get_empty_fib_id()` returns error.